File: src/content/3.8/code/haskell/snippet06.hs
evalZ :: Expr -> Integer
evalZ RZero        = 0
evalZ ROne         = 1
evalZ (RAdd e1 e2) = evalZ e1 + evalZ e2
evalZ (RMul e1 e2) = evalZ e1 * evalZ e2
evalZ (RNeg e)     = -(evalZ e)

File: src/content/3.8/code/scala/snippet06.scala
def evalZ: Expr => Int = {
  case RZero => 0
  case ROne => 1
  case RAdd(e1, e2) => evalZ(e1) + evalZ(e2)
  case RMul(e1, e2) => evalZ(e1) * evalZ(e2)
  case RNeg(e) => -evalZ(e)
}